internal/cluster/cluster.go

```diff
@@ -211,6 +216,64 @@ func (m *Manager) DeregisterCluster(ctx context.Context, c *common.Cluster) erro
 	return err
 }
 
+func (m *Manager) reportValidationFailedMetrics(ctx context.Context, c *common.Cluster, ocpVersion, emailDomain string) error {
+	log := logutil.FromContext(ctx, m.log)
+	var validationRes validationsStatus
+	if c.ValidationsInfo != "" {
+		if err := json.Unmarshal([]byte(c.ValidationsInfo), &validationRes); err != nil {
+			log.WithError(err).Errorf("Failed to unmarshal validations info from cluster %s", *c.ID)
+			return err
+		}
+	}
+	for _, vRes := range validationRes {
+		for _, v := range vRes {
+			if v.Status == ValidationFailure {
+				m.metricAPI.ClusterValidationFailed(ocpVersion, emailDomain, models.ClusterValidationID(v.ID))
+			}
+		}
+	}
+	return nil
+}
+
+func (m *Manager) reportValidationStatusChanged(ctx context.Context, c *common.Cluster,
+	newValidationRes, currentValidationRes validationsStatus) {
+	for vCategory, vRes := range currentValidationRes {
+		for i, v := range vRes {
+			if newValidationRes[vCategory][i].Status == ValidationFailure && v.Status == ValidationSuccess {
+				m.metricAPI.ClusterValidationChanged(c.OpenshiftVersion, c.EmailDomain, models.ClusterValidationID(v.ID))
+				eventMsg := fmt.Sprintf("Cluster validation '%s' that used to succeed is now failing", v.ID)
+				m.eventsHandler.AddEvent(ctx, *c.ID, nil, models.EventSeverityWarning, eventMsg, time.Now())
+			}
+			if newValidationRes[vCategory][i].Status == ValidationSuccess && v.Status == ValidationFailure {
+				eventMsg := fmt.Sprintf("Cluster validation '%s' is now fixed", v.ID)
+				m.eventsHandler.AddEvent(ctx, *c.ID, nil, models.EventSeverityInfo, eventMsg, time.Now())
+			}
+		}
+	}
+}
+
+func (m *Manager) checkValidationChanged(ctx context.Context, c *common.Cluster, newValidationRes validationsStatus) (validationsStatus, bool, error) {
+	var currentValidationRes validationsStatus
+	if c.ValidationsInfo != "" {
+		if err := json.Unmarshal([]byte(c.ValidationsInfo), &currentValidationRes); err != nil {
+			return validationsStatus{}, false, errors.Wrapf(err, "Failed to unmarshal validations info from cluster %s", *c.ID)
+		}
+	}
+	if len(newValidationRes) == 0 {
```

Comment:
This is strange.
User changing the cluster configuration (Setting `userManageNetwork` for example) might reduce the number of validations the cluster need to pass.
Why do need this if?

Reply:
Ummm does it means the validation won't be part of the json string at all or that we will just allow `failure` status on it?
The reason it is there is because otherwise we will hit a panic when trying to access and "out of bound" array index.
We must ensure that the validations_info struct don't contains less fields then it previously did (regardless of the status result)

Reply:
See my last comment here:
https://github.com/openshift/assisted-service/pull/1198/files#r596672595